### ISSUE TITLE
Pyright incorrectly determines attribute in subclass overriding generic attribute in parent as type Never

### ISSUE THREAD

**Issue by robert-bn** [microsoft/pyright#8672]
Consider the following code:
```python
from typing import assert_never

class A[T]:
    value: T

class AInt(A[int]):
    value: int

def f[T](x: A[T]):
    if isinstance(x, AInt):
        assert_never(x.value)
```
pyright 1.1.373 correctly determines that the type of `x` is int, and reports the error:
```
error: Argument of type "int" cannot be assigned to parameter "arg" of type "Never" in function "assert_never"
    Type "int" is incompatible with type "Never" (reportArgumentType)
```
but pyright 1.1.374 reports no error.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be addressed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.375.

**Comment by AdrianB-sovo** [comment]
I'm have the same kind of problem with the following code today:
```python
from typing import assert_never

for i, item in enumerate(["a", "b", "c"]):
    match i:
        case 0:
            test = item
        case 1:
            test = item
        case 2:
            test = item
        case _:
            assert_never(i)
```

```
Argument of type "int" cannot be assigned to parameter "arg" of type "Never" in function "assert_never"
  Type "int" is not assignable to type "Never"Pylance[reportArgumentType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportArgumentType)
```

In the Python Language Server output, I have the following:
```Pylance language server 2024.9.2 (pyright version 1.1.378, commit ce64adc1) starting```

**Comment by erictraut** [comment]
@AdrianB-sovo, your issue isn't related to this bug report. In your code sample, `enumerate` returns a value whose `__next__` method is defined as returning a value of type `tuple[int, str]`. That means `i` has a type of `int`. A static type checker does not have enough information to narrow this type to `Never` in the match statement. So pyright is working as intended here. If you have further questions about this case, feel free to post to the discussion section.
